rp2040-hal/src/dma/mod.rs

```diff
@@ -0,0 +1,265 @@
+        }
+
+        /// Set of DMA channels.
+        pub struct Channels {
+            $(
+                /// DMA channel.
+                pub $chX: Channel<$CHX>,
+            )+
+        }
+        $(
+            /// DMA channel identifier.
+            pub struct $CHX;
+            impl ChannelIndex for $CHX {
+                fn id() -> u8 {
+                    $x
+                }
+            }
+        )+
+
+        /// Set of DMA channels.
+        pub struct DynChannels {
+            $(
+                /// DMA channel.
+                pub $chX: Option<Channel<$CHX>>,
+            )+
+        }
+    }
+}
+
+channels! {
+    CH0: (ch0, 0),
+    CH1: (ch1, 1),
+    CH2: (ch2, 2),
+    CH3: (ch3, 3),
+    CH4: (ch4, 4),
+    CH5: (ch5, 5),
+    CH6: (ch6, 6),
+    CH7: (ch7, 7),
+    CH8: (ch8, 8),
+    CH9: (ch9, 9),
+    CH10:(ch10, 10),
+    CH11:(ch11, 11),
+}
+
+trait ChannelRegs {
+    unsafe fn ptr() -> *const rp2040_pac::dma::CH;
+    fn regs(&self) -> &rp2040_pac::dma::CH;
+}
+
+impl<CH: ChannelIndex> ChannelRegs for Channel<CH> {
+    unsafe fn ptr() -> *const rp2040_pac::dma::CH {
+        &(*rp2040_pac::DMA::ptr()).ch[CH::id() as usize] as *const _
+    }
+
+    fn regs(&self) -> &rp2040_pac::dma::CH {
+        unsafe { &*Self::ptr() }
+    }
+}
+
+/// Trait which is implemented by anything that can be read via DMA.
+pub trait ReadTarget {
+    /// Type which is transferred in a single DMA transfer.
+    type ReceivedWord;
+
+    /// Returns the DREQ number for this data source (`None` for memory buffers).
+    fn rx_treq() -> Option<u8>;
+
+    /// Returns the address and the maximum number of words that can be transferred from this data
+    /// source in a single DMA operation.
+    ///
+    /// For peripherals, the count should likely be u32::MAX. If a data source implements
+    /// EndlessReadTarget, it is suitable for infinite transfers from or to ring buffers. Note that
+    /// ring buffers designated for endless transfers, but with a finite buffer size, should return
+    /// the size of their individual buffers here.
+    ///
+    /// # Safety
+    ///
+    /// This function has the same safety guarantees as `StaticReadBuffer::static_read_buffer`.
+    fn rx_address_count(&self) -> (u32, u32);
+
+    /// Returns whether the address shall be incremented after each transfer.
+    fn rx_increment(&self) -> bool;
+}
+
+/// Marker which signals that `rx_address_count()` can be called multiple times.
+///
+/// The DMA code will never call `rx_address_count()` to request more than two buffers to configure
+/// two DMA channels. In the case of peripherals, the function can always return the same values.
+pub trait EndlessReadTarget: ReadTarget {}
+
+impl<B: ReadBuffer> ReadTarget for B {
+    type ReceivedWord = <B as ReadBuffer>::Word;
+
+    fn rx_treq() -> Option<u8> {
+        None
+    }
+
+    fn rx_address_count(&self) -> (u32, u32) {
+        // Safety: We only call the function once per buffer.
```

Comment:
I don't understand this safety comment: As `rx_address_count` is not unsafe, what keeps some random safe code from calling this method multiple times?

Reply:
Comment was from the original PR, but agree that it doesn't make sense.
Maybe the original wasn't going to expose this to end-users?
Also not sure what the side-effect is that was hoped to be avoided, I think I'll write some tests (after merge) to exercise this condition to see if it's a problem.

Reply:
ah.
```rust
/// # Safety
///
/// Once this method has been called, it is unsafe to call any `&mut self`
/// methods on this object as long as the returned value is in use (by DMA).
unsafe fn read_buffer(&self) -> (*const Self::Word, usize);
```
but this is also true if the addresses returned are copied and used for another DMA transfer - the function in and of itself is safe. I'll remove the comment as it's more confusing than helpful.